adoptingerlang.org

```diff
@@ -1334,6 +1334,370 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
+
+You should, therefore, aim to keep the lock file as long as you can and to change it only as required. You can do so by using the following operations:
+
+- =rebar3 unlock <appname>=, which removes an unused dependency from the lockfile. You will usually want to call this _after_ you have already removed it from your =rebar.config= file, to tell rebar3 that it's really gone or has been downgraded to a transitive dependency
+- =rebar3 upgrade <appname>=, which tells rebar3 to disregard the locked version for that application, and re-build the dependency tree from the version currently specified in =rebar.config= (if any). This will generate a new lock file and re-resolve all transitive dependencies that might have changed.
+- =rebar3 update=, while not strictly about the lock file, updates the local _snapshot_ of remote hex packages, a kind of cache that prevents each build from pinging the package server. If you find yourself calling =rebar3 upgrade= on an app and it isn't upgrading to the latest version you know is available in Hex, you will want to =update= first. This is because rebar3 limits the use of the network by trying to resolve dependencies with the local index cache. =rebar3 update= will fetch the latest index entries for each package already in the index and then another run of =upgrade= will see the latest versions. Note that if you specify an exact version to upgrade to rebar3 will automatically fetch the updated index because it is unable to satisfy the dependency locally.
+- =rebar3 tree= will print out the dependency tree that was built and is represented by the lock file
+- =rebar3 deps= will list out dependencies and annotate those that could be updated. Do note that there are strong limitations in what it considers worth updating: a branch in git, a tag whose reference shifted, or hex versions that weren't specified. But if you specified a package to have version ="1.2.3"= and =1.2.4= is available, it won't tell you anything
+
+There are other commands that are a bit more drastic, namely =rebar3 unlock= and =rebar3 upgrade= (without any arguments). These will just get rid of the lock file for the next build of the project. But overall, all these commands will do everything you need to manage most dependencies.
+
+In general your workflow might look like this:
+
+1. set up the initial project, compile once, and track the lock file in source control
+2. find out you have a top-level dependency you want to change
+3. change the dependency definition in the rebar.config file (or optionally if using a non-versioned hex package, call =rebar3 update=)
+4. call =rebar3 upgrade <app>= to update the application and its transitive dependencies
+
+And that's it, you're done.
+
+*** Checkout Dependencies
+
```

Comment:
```suggestion

Rebar3 wants to make the developers life easier while also staying safe and repeatable. =_checkouts= are a feature for keeping a quality of life task that goes against the lock file and repeatability, quick feedback look on local changes to dependencies, simple and controlled.

```

Good to introduce that the feature is for local dependencies quick I think, and to setup for the fact that it will result in the dep being removed from the lock file. I don't think this intro is done but is roughly what I had in mind.